rust/kernel/error.rs

```diff
@@ -5,60 +5,77 @@
 //! C header: [`include/uapi/asm-generic/errno-base.h`](../../../include/uapi/asm-generic/errno-base.h)
 
 use crate::str::CStr;
-use crate::{bindings, c_types};
+use crate::{bindings, c_types, static_assert};
 use alloc::{alloc::AllocError, collections::TryReserveError};
 use core::convert::From;
 use core::fmt;
-use core::num::TryFromIntError;
+use core::num::{NonZeroI32, TryFromIntError};
 use core::str::{self, Utf8Error};
 
 /// Generic integer kernel error.
 ///
 /// The kernel defines a set of integer generic error codes based on C and
 /// POSIX ones. These codes may have a more specific meaning in some contexts.
 ///
+/// We use NonZeroI32 as the inner type because Result<(), Error> can fit in
+/// 32 bit memory, and it takes one register (rather than 2 for i16 or i32)
+/// for returning or argument passing which is fast in many architectures.
+///
 /// # Invariants
 ///
 /// The value is a valid `errno` (i.e. `>= -MAX_ERRNO && < 0`).
 #[derive(Clone, Copy, PartialEq, Eq)]
-pub struct Error(c_types::c_int);
+pub struct Error(NonZeroI32);
+
+/// Converts kernel const to Error.
+/// Note: The value stored in Error is negative
+/// DO NOT use this macro outside of Error const initializations.
+macro_rules! kernel_const_to_error {
+    ($($tt:tt)*) => {
+        {
+            static_assert!(-($($tt)* as i32) >= -(bindings::MAX_ERRNO as i32)
+            && -($($tt)* as i32) < 0);
+            Error(unsafe { NonZeroI32::new_unchecked(-($($tt)* as i32)) })
+        }
+    };
+}
 
 impl Error {
     /// Invalid argument.
-    pub const EINVAL: Self = Error(-(bindings::EINVAL as i32));
+    pub const EINVAL: Self = kernel_const_to_error!(bindings::EINVAL);
```

Comment:
If we are going to use a macro for this, should we do everything at once? i.e. something like:

```rust
define_const_error!(EINVAL, "Invalid argument");
```

Reply:
Never mind, you were discussing this above. Resolving...